### PR TITLE
Move out parsing into perception

### PR DESCRIPTION
**Pull request by kavyasrinet**
# Description

This PR moves the actual parsing (text -> logical form) or (text -> text reply ) to perception and disentangles 
the grounding of parses to dialogue objects.

Changeset here: #456 
Stage 2 summarized here: #337 

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [x] I want a deep design review.

## Before and After

This change enables us to get closer to the pseudo code here: https://gist.github.com/soumith/787a51d0ee6e01e910d0e2b8cd56b159

TL;DR:
- We now get parse and or reply separately (disentangled from associating them with DialogueObjects as before) from the datasets and models.
- `init_perception` initializes this perception mode
- `dialogue_manager.step()` now takes in the parse / reply and takes care of associating them with respective DialogueObjects and pushing them to `dialogue_stack` (compared to taking text before).
- Cleaning up of `dialog` folder and moving relevant model and dataset files to `perception`

# Testing

Tested locally on devfair and on CI

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

**Comment by kavyasrinet** [comment]
Review guidelines:
1. For the first commit that's pretty big, please review the files : 
- `agents/craftassist/craftassist_agent.py`
- `agents/loco_mc_agent.py`
- `agents/locobot/locobot_agent.py`
- `droidlet/dialog/dialogue_manager.py`
- `droidlet/dialog/parse_to_dialogue_object.py`
- `droidlet/perception/semantic_parsing_model/droidlet_nsp_model_wrapper.py`
- `droidlet/perception/semantic_parsing_model/load_and_check_datasets.py`
- The folder : `droidlet/perception/semantic_parsing_model/`. Nit : this has been cleaned up and refactored in later commits. But high level refactor is here, so ignore some empty files there.
2. Some open questions:
- Should we set the reply / logical form in memory somewhere explicitly ? (the output of `chat_parser.get_parse(chat)`
- Do we need to still use the `get_last_m_chats` from `DialogueManager` in DialogueObjectMapper. We were using this before since this class was taking in chat and getting, parse, pushing DialogueObjects, there doesn't seem to be a need anymore since the `DialogueManager`'s step function already gets the chat and parse.
 - Remaining nit: move chat_parser in init_perception to perception modules

File: agents/loco_mc_agent.py [hunk 1]
```diff
@@ -284,14 +281,15 @@ def controller_step(self):
                 self.perceive(force=True)
             # change this to memory.get_time() format?
             self.last_chat_time = time.time()
+            # For now just process the first incoming chat, where chat -> [speaker, chat]
+            chat_parse = self.chat_parser.get_parse(incoming_chats[0][1])
```
Comment (aszlam) [inline agents/loco_mc_agent.py:285]: sorry, I don't understand, shouldn't this, and the whole block, be in the .perceive()?

File: droidlet/memory/sql_memory.py [hunk 2]
```diff
@@ -578,6 +579,16 @@ def get_chat_by_id(self, memid: str) -> "ChatNode":
         """
         return ChatNode(self, memid)
 
+    def get_chat_id(self, speaker_id: str, chat: str) -> str:
```
Comment (aszlam) [inline droidlet/memory/sql_memory.py:582]: I don't mind this here rn, but I am probably going to torch it in a few commits with the updated search syntax

File: droidlet/memory/sql_memory.py [hunk 3]
```diff
@@ -609,6 +620,26 @@ def get_most_recent_incoming_chat(self, after=-1) -> Optional["ChatNode"]:
         else:
             return None
 
+    ###################
+    ## Logical form ###
+    ###################
+
+    def add_logical_form(self, logical_form: dict):
```
Comment (aszlam) [inline droidlet/memory/sql_memory.py:627]: similarly... I think we should be moving just using the ProgramNode.create

File: droidlet/memory/sql_memory.py [hunk 4]
```diff
@@ -609,6 +620,26 @@ def get_most_recent_incoming_chat(self, after=-1) -> Optional["ChatNode"]:
...
+        """Create a new ProgramNode
+
+        Args:
+            logical_form: the semantic parser's output
+        """
+        return ProgramNode.create(self, logical_form)
+
+    def get_logical_form_by_id(self, memid: str) -> "ProgramNode":
```
Comment (aszlam) [inline droidlet/memory/sql_memory.py:635]: this also... new search syntax will make these extraneous

File: droidlet/memory/sql_memory.py (hunk 2 above)
Comment (kavyasrinet) [reply droidlet/memory/sql_memory.py:582]: But you are okay with this in this PR ?

File: droidlet/memory/sql_memory.py (hunk 3 above)
Comment (kavyasrinet) [reply droidlet/memory/sql_memory.py:627]: same as above ^

File: droidlet/memory/sql_memory.py (hunk 4 above)
Comment (kavyasrinet) [reply droidlet/memory/sql_memory.py:635]: same as above ^

File: agents/loco_mc_agent.py [hunk 5]
```diff
@@ -276,22 +269,36 @@ def controller_step(self):
...
-            # to here ###########################################
-            # for now just process the first incoming chat
-            self.dialogue_manager.step(incoming_chats[0])
-        else:
+            # For now just process the first incoming chat, where chat -> [speaker, chat]
+            speaker, chat = incoming_chats[0]
+            preprocessed_chat, chat_parse = self.chat_parser.get_parse(chat)
+            # add postprocessed chat here
+            chat_memid = self.memory.add_chat(self.memory.get_player_by_name(speaker).memid, preprocessed_chat)
+            logical_form_memid = self.memory.add_logical_form(chat_parse)
```
Comment (aszlam) [inline agents/loco_mc_agent.py:283]: as mentioned in the sql_memory.py comment, I think its better to do a ProgramNode.create(), but I don't have strong opinion.  longer term, do you want to keep the main memory creation?  we should do one or the other; and if we standardize on main memory then I think we should have a more generic creation interface

File: agents/loco_mc_agent.py [hunk 6]
```diff
@@ -276,22 +269,36 @@ def controller_step(self):
...
+            self.memory.add_triple(subj=chat_memid, pred_text="has_logical_form", obj=logical_form_memid)
+            # TODO: tag as unprocessed
+            self.memory.tag(subj_memid=chat_memid, tag_text="unprocessed")
+
+        for v in self.perception_modules.values():
+            v.perceive(force=force)
+
+
+
+    def controller_step(self):
```
Comment (aszlam) [inline agents/loco_mc_agent.py:293]: nice!

File: droidlet/memory/sql_memory.py (hunk 2 above)
Comment (aszlam) [reply droidlet/memory/sql_memory.py:582]: yes, its fine in this PR

File: droidlet/memory/sql_memory.py (hunk 3 above)
Comment (aszlam) [reply droidlet/memory/sql_memory.py:627]: also fine

File: droidlet/memory/sql_memory.py (hunk 4 above)
Comment (aszlam) [reply droidlet/memory/sql_memory.py:635]: also fine- commenting to warn rather than asking for change

**Review by aszlam — APPROVED** [review]

File: agents/loco_mc_agent.py (hunk 5 above)
Comment (kavyasrinet) [reply agents/loco_mc_agent.py:283]: Yeah I  have a preference for keeping these as methods / APIs exposed from memory instead of directly accessing from memory nodes, where possible.
